Implement instantiating and unboxing through portable stublinker code

davidwrighton opened the pull request:
- Handle only the cases with register to register moves
- Followon work to remove the old instantiating/unboxing stub infra (except for x86, x86 will remain special)
- Shares abi processing logic with delegate shuffle thunk creation
- Architecture specific logic is relatively simple

Performance impact on startup for Windows x64 is minimal
- Example test applications such as powershell and webapi template now compile 3-4 instantiation stubs where in the past they were able to stublinker generate all of them.

Throughput improvement for Windows x64 is visible in microbenchmarks, but not considered to be important.

Linux x64 now compiles effectively 0 instantiating stubs in any normal scenario. 
- Substantial count of instantiation stubs jitting removed

Also, code quality is higher than the jit as these stublinker generated stubs will reliably tail call, and do not need gc polls.

davidwrighton committed:
Implement instantiating and unboxing through portable stublinker code
- Handle only the cases with register to register moves
- Followon work to remove the old instantiating/unboxing stub infra (except for x86, x86 will remain special)
- Shares abi processing logic with delegate shuffle thunk creation
- Architecture specific logic is relatively simple

davidwrighton committed:
Fix cross arch compile issues

davidwrighton committed:
oops

davidwrighton committed:
Fix bugs on Amd64

davidwrighton committed:
Fix Windows build, and ADM64 asserts

davidwrighton committed:
Fix GetArgLoc for stack structure arguments on Windows Amd64

davidwrighton committed:
Make x86 build, and remove excess unneeded code for non-il stub based invokes and thunks

davidwrighton committed:
Remove architecture specific logic around secure delegates- All architectures use IL stubs for secure delegates- No real impact as only arm32 uses the secure delegate infra at allUse FEATURE_INSTANTIATINGSTUB_AS_IL to indicate IL stub based unboxing and instantiating stubsUse FEATURE_PORTABLE_SHUFFLE_THUNKS to indicate the generally portable unboxing/instantiating stubs for simple cases

davidwrighton committed:
Merge branch 'master' of https://github.com/dotnet/coreclr into architecture_neutral_assemblyinstantiationstubgen

davidwrighton committed:
Merge branch 'master' of https://github.com/dotnet/coreclr into architecture_neutral_assemblyinstantiationstubgen

davidwrighton committed:
Fix merge error

davidwrighton committed:
Fix issues noted in testing
- Fix arm32 build
- Windows x64 may use fp registers to pass arguments
  - Tweak GetArgLoc to work for that case
  - Do not permit use of HELPERREG in computed instantiating stubs

davidwrighton committed:
EmitComputeInstantiatingMethodStub for arm/arm64 platforms does not use FP or helper registers

davidwrighton committed:
Fix re-arranging of args in shuffle thunks
- When floating point args were in use, it overran the pGraphNodes buffer
- Fix by adding the NUM_FLOAT_ARGUMENT_REGISTERS define to all architectures that use floating point registers

davidwrighton committed:
Stress test for stub generation

davidwrighton committed:
Fix typo

davidwrighton commented:
/azp run coreclr-outerloop

maryamariyan commented:
Thank you for your contribution. As announced in dotnet/coreclr#27549 this repository will be moving to dotnet/runtime on November 13. If you would like to continue working on this PR after this date, the easiest way to move the change to dotnet/runtime is:
1.	In your coreclr repository clone, create patch by running `git format-patch origin`
2.	In your runtime repository clone, apply the patch by running `git apply --directory src/coreclr <path to the patch created in step 1>`

jkotas commented on src/vm/dllimport.h:
```diff
@@ -226,8 +226,8 @@ inline bool SF_IsArrayOpStub           (DWORD dwStubFlags) { LIMITED_METHOD_CONT
 #endif
 
-#ifdef FEATURE_STUBS_AS_IL
 inline bool SF_IsSecureDelegateStub  (DWORD dwStubFlags)    { LIMITED_METHOD_CONTRACT; return (dwStubFlags == ILSTUB_SECUREDELEGATE_INVOKE); }
```
Maybe also rename SecureDelegate to something more sensible while you are on it?

davidwrighton replied:
They are now WrapperDelegate, and I removed the last vestige of "creator" assembly from the data structures.

davidwrighton committed:
Adjust stubs abi test count to 100 to reduce test load

davidwrighton committed:
Merge branch 'master' of https://github.com/dotnet/coreclr into architecture_neutral_assemblyinstantiationstubgen

davidwrighton committed:
Rename secure delegates to wrapper delegates
- Secure delegates are no longer a feature of the runtime
- But the wrapper delegate lives on as a workaround for a weird detail of the ARM32 abi

davidwrighton committed:
Merge branch 'master' of https://github.com/dotnet/coreclr into architecture_neutral_assemblyinstantiationstubgen

maryamariyan commented:
Thank you for your contribution. As announced in #27549 the dotnet/runtime repository will be used going forward for changes to this code base. Closing this PR as no more changes will be accepted into master for this repository. If you’d like to continue working on this change please move it to dotnet/runtime.
